core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigurationDownloader.java

```diff
@@ -0,0 +1,101 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import com.google.gson.stream.JsonReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.Map;
+import org.sonarqube.ws.QualityProfiles.SearchWsResponse.QualityProfile;
+import org.sonarqube.ws.client.WsResponse;
+import org.sonarsource.sonarlint.core.container.connected.SonarLintWsClient;
+import org.sonarsource.sonarlint.core.proto.Sonarlint.GlobalProperties;
+import org.sonarsource.sonarlint.core.proto.Sonarlint.ModuleConfiguration;
+import org.sonarsource.sonarlint.core.util.StringUtils;
+
+public class ModuleConfigurationDownloader {
+
+  private final SonarLintWsClient wsClient;
+  private final ModuleHierarchyDownloader moduleHierarchyDownloader;
+  private final ModuleQualityProfilesDownloader moduleQualityProfilesDownloader;
+
+  public ModuleConfigurationDownloader(SonarLintWsClient wsClient, ModuleHierarchyDownloader moduleHierarchyDownloader,
+    ModuleQualityProfilesDownloader moduleQualityProfilesDownloader) {
+    this.wsClient = wsClient;
+    this.moduleHierarchyDownloader = moduleHierarchyDownloader;
+    this.moduleQualityProfilesDownloader = moduleQualityProfilesDownloader;
+  }
+
+  public ModuleConfiguration fetchModuleConfiguration(String moduleKey, GlobalProperties globalProps) {
+    ModuleConfiguration.Builder builder = ModuleConfiguration.newBuilder();
+    fetchProjectQualityProfiles(moduleKey, builder);
+    fetchProjectProperties(moduleKey, globalProps, builder);
+    fetchModuleHierarchy(moduleKey, builder);
+
+    return builder.build();
+  }
+
+  private void fetchModuleHierarchy(String moduleKey, ModuleConfiguration.Builder builder) {
+    Map<String, String> moduleHierarchy = moduleHierarchyDownloader.fetchModuleHierarchy(moduleKey);
+    builder.putAllModulePathByKey(moduleHierarchy);
+  }
+
+  private void fetchProjectQualityProfiles(String moduleKey, ModuleConfiguration.Builder builder) {
+    for (QualityProfile qp : moduleQualityProfilesDownloader.fetchModuleQualityProfiles(moduleKey)) {
+      builder.putQprofilePerLanguage(qp.getLanguage(), qp.getKey());
+    }
+  }
+
+  private void fetchProjectProperties(String moduleKey, GlobalProperties globalProps, ModuleConfiguration.Builder projectConfigurationBuilder) {
+    WsResponse response = wsClient.get("/api/properties?format=json&resource=" + StringUtils.urlEncode(moduleKey));
```

Comment:
This could be extracted to a class (a ProjectPropertiesDownloader) to be more easilly tested and maintained